zowe-rest-api-sample-spring/docs/zos-security.md

```diff
@@ -199,3 +279,45 @@ The shared library `libsecur.so` is expected to be in zFS filesystem in the a di
+
+The REST API service that is running on z/OS typically needs to check if the user that is authenticated has access to specific SAF resources to protect access to the functionality of the REST API or validate that the user can do the action before the action is started.
+
+The API to check access to SAF resource is a part of `PlatformSecurityService` interface.
+
+It provides following methods:
+
+- `boolean checkPermission(String userid, String resourceClass, String resourceName, AccessLevel accessLevel, boolean resourceHasToExist)`
+- `boolean checkPermission(String userid, String resourceClass, String resourceName, AccessLevel accessLevel)`
+- `boolean checkPermission(String resourceClass, String resourceName, AccessLevel accessLevel, boolean resourceHasToExist)`
+- `boolean checkPermission(String resourceClass, String resourceName, AccessLevel accessLevel)`
+
+All of them return `true` if user has access to the resource and `false` if the user has not access to the resource or the resource does not exist. They throw `AccessControlError` in other cases of failures (invalid user ID, invalid resource name, internal SAF error, misconfiguration of the service). The optional parameter `resourceHasToExist` is by default `true` and it causes that non-existing resource means that that no user has access to it. Some applications may want the opposite, protect access only when the security administator has created the resouces, then they can use `false` as the value of `resourceHasToExist` parameter.
+
+If the Spring profile `zos` is active, then this provider uses `SafPlatformAccessControl` class that uses Java Reflection to
+call `com.ibm.os390.security.PlatformUser` class. This class is available in IBM® SDK for z/OS®, Java™ Technology Edition in `racf.jar`.
+This JAR is not available in public repositories and cannot be added to this repository. These APIs are documented in <https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.zsecurity.80.doc/zsecurity-component/saf.html>.
+
+When you run it outside of z/OS without `zos` profile, a mock implementation `MockPlatformAccessControl` is used. This has predefined values that are accepted.
+
+### Protecting access to REST API endpoints
+
+The REST API endpoints can be protected by the `org.springframework.security.access.prepost.PreAuthorize` annotation.
+
+The SDK defined two new securiry expressions:
+
+- `boolean hasSafResourceAccess(String resourceClass, String resourceName, String accessLevel)` - return `true` when user has access to the resource
+- `boolean hasSafServiceResourceAccess(String resourceNameSuffix, String accessLevel)` - similar as the previous one but the resource class and resource name prefix is taken from the service configuration in `application.yml` under key `zowe.commons.security.saf`
+
+So you can do following:
+
+```java
+@GetMapping("/safProtectedResource")
+@PreAuthorize("hasSafResourceAccess('FACILITY', 'BPX.SERVER', 'UPDATE')")
```

Comment:
this is a really nice implementation!

Reply:
Thanks! I was considering to use existing Spring Security expression `hasAuthority()` but I was not sure how to put class, name, and level into one string without making it understandable so I end up creating a new expression.
https://docs.spring.io/autorepo/docs/spring-security/4.0.0.RC1/reference/html/el-access.html